File: repos/banana-fantasy/utils/slowDraftClock.ts
/**
 * Slow-draft pick clock: only 05:00–22:00 America/Los_Angeles counts (22:00–05:00 PT paused).
 * Mirrors sbs-drafts-api/models/slow_draft_clock.go — keep in sync.
 */

const PACIFIC = 'America/Los_Angeles';
const PAUSE_START_HOUR = 22;

/**
 * PT hour the overnight pause ends. Legacy 5. The SlowClockProvider sets it
 * from system_config/slowDraftClock.pauseEndHour when the switch is active
 * (Richard 8/26: 7). Every helper takes it as an optional trailing arg and
 * defaults to this module value, so callers don't have to thread it through.
 */
let defaultPauseEndHour = 5;
export function setSlowDraftPauseEndHour(h: number): void {
  if (Number.isInteger(h) && h > 0 && h < PAUSE_START_HOUR) defaultPauseEndHour = h;
}
export function getSlowDraftPauseEndHour(): number {
  return defaultPauseEndHour;
}

type NYWall = { y: number; mon: number; d: number; sod: number };

function nyWallFromUnixSec(unixSec: number): NYWall {
  const ms = unixSec * 1000;
  const parts = new Intl.DateTimeFormat('en-US', {
    timeZone: PACIFIC,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  }).formatToParts(new Date(ms));
  const map: Record<string, string> = {};
  for (const p of parts) {
    if (p.type !== 'literal') {
      map[p.type] = p.value;
    }
  }
  const y = Number(map.year);
  const mon = Number(map.month);
  const d = Number(map.day);
  // Intl returns '24' instead of '00' for midnight in 24-hour mode; normalize.
  const h = Number(map.hour) % 24;
  const min = Number(map.minute);
  const s = Number(map.second);
  const sod = h * 3600 + min * 60 + s;
  return { y, mon, d, sod };
}

function compareNYWall(a: NYWall, y: number, mon: number, d: number, sod: number): number {
  if (a.y !== y) return a.y - y;
  if (a.mon !== mon) return a.mon - mon;
  if (a.d !== d) return a.d - d;
  return a.sod - sod;
}

function unixAtNYWallClock(
  y: number,
  mon: number,
  d: number,
  hour: number,
  minute: number,
  second: number
): number {
  const targetSod = hour * 3600 + minute * 60 + second;
  const noonUtcGuess = Math.floor(Date.UTC(y, mon - 1, d, 12, 0, 0) / 1000);
  let lo = noonUtcGuess - 20 * 3600;
  let hi = noonUtcGuess + 20 * 3600;
  let ans = -1;
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2);
    const w = nyWallFromUnixSec(mid);
    const c = compareNYWall(w, y, mon, d, targetSod);
    if (c < 0) {
      lo = mid + 1;
    } else {
      ans = mid;
      hi = mid - 1;
    }
  }
  if (ans === -1) {
    throw new Error(`unixAtNYWallClock: no match for ${y}-${mon}-${d} ${hour}:${minute}:${second}`);
  }
  return ans;
}

function addCalendarDays(
  y: number,
  mon: number,
  d: number,
  deltaDays: number
): [number, number, number] {
  const dt = new Date(Date.UTC(y, mon - 1, d + deltaDays));
  return [dt.getUTCFullYear(), dt.getUTCMonth() + 1, dt.getUTCDate()];
}

function inNightPause(w: NYWall, pauseEndHour: number): boolean {
  return w.sod >= PAUSE_START_HOUR * 3600 || w.sod < pauseEndHour * 3600;
}

function advanceToNextActiveUnix(unixSec: number, pauseEndHour: number): number {
  const w = nyWallFromUnixSec(unixSec);
  if (!inNightPause(w, pauseEndHour)) return unixSec;
  if (w.sod >= PAUSE_START_HOUR * 3600) {
    const [y2, m2, d2] = addCalendarDays(w.y, w.mon, w.d, 1);
    return unixAtNYWallClock(y2, m2, d2, pauseEndHour, 0, 0);
  }
  return unixAtNYWallClock(w.y, w.mon, w.d, pauseEndHour, 0, 0);
}


/**
 * `freshAfterPause` (system_config/slowDraftClock.freshClockAfterPause): a pick
 * that would cross the 22:00 PT pause does NOT carry its leftover minutes into
 * the morning — it restarts with the FULL pickLengthSec at 05:00 PT. Default
 * false = legacy carry-over. Mirrors slowDraftPickEndUnixOpts in Go.
 */
export function slowDraftPickEndUnix(fromUnix: number, pickLengthSec: number, freshAfterPause = false, pauseEndHour: number = defaultPauseEndHour): number {
  if (pickLengthSec <= 0) return fromUnix;
  // Longest pick that still fits one active window — beyond it a fresh clock could never end.
  if (pickLengthSec > (PAUSE_START_HOUR - pauseEndHour) * 3600) freshAfterPause = false;
  let cur = advanceToNextActiveUnix(fromUnix, pauseEndHour);
  let remaining = pickLengthSec;
  while (remaining > 0) {
    const w = nyWallFromUnixSec(cur);
    const windowClose = unixAtNYWallClock(w.y, w.mon, w.d, PAUSE_START_HOUR, 0, 0);
    const avail = windowClose - cur;
    if (avail <= 0) {
      const [y2, m2, d2] = addCalendarDays(w.y, w.mon, w.d, 1);
      cur = unixAtNYWallClock(y2, m2, d2, pauseEndHour, 0, 0);
      continue;
    }
    if (remaining <= avail) {
      return cur + remaining;
    }
    remaining = freshAfterPause ? pickLengthSec : remaining - avail;
    const [y3, m3, d3] = addCalendarDays(w.y, w.mon, w.d, 1);
    cur = unixAtNYWallClock(y3, m3, d3, pauseEndHour, 0, 0);
  }
  return cur;
}

export function slowDraftEffectiveElapsedSeconds(startUnix: number, endUnix: number, pauseEndHour: number = defaultPauseEndHour): number {
  if (endUnix <= startUnix) return 0;
  let total = 0;
  let cur = advanceToNextActiveUnix(startUnix, pauseEndHour);
  while (cur < endUnix) {
    const w = nyWallFromUnixSec(cur);
    const windowClose = unixAtNYWallClock(w.y, w.mon, w.d, PAUSE_START_HOUR, 0, 0);
    if (windowClose <= cur) {
      const [y2, m2, d2] = addCalendarDays(w.y, w.mon, w.d, 1);
      cur = unixAtNYWallClock(y2, m2, d2, pauseEndHour, 0, 0);
      continue;
    }
    let chunkEnd = windowClose;
    if (endUnix < chunkEnd) chunkEnd = endUnix;
    total += chunkEnd - cur;
    if (chunkEnd >= endUnix) break;
    const [y3, m3, d3] = addCalendarDays(w.y, w.mon, w.d, 1);
    cur = unixAtNYWallClock(y3, m3, d3, pauseEndHour, 0, 0);
  }
  return total;
}

export function slowDraftDisplayedSecondsRemaining(
  nowUnixSec: number,
  pickStartUnixSec: number,
  pickLengthSec: number,
  pauseEndHour: number = defaultPauseEndHour
): number {
  const elapsed = slowDraftEffectiveElapsedSeconds(pickStartUnixSec, nowUnixSec, pauseEndHour);
  return Math.max(0, Math.floor(pickLengthSec - elapsed));
}

export function isSlowDraftPickLength(pickLengthSec: number): boolean {
  return pickLengthSec >= 3600;
}

/** True if `nowUnixSec` falls inside the slow-draft night pause (22:00–05:00 PT). */
export function isSlowDraftNightPause(nowUnixSec: number, pauseEndHour: number = defaultPauseEndHour): boolean {
  return inNightPause(nyWallFromUnixSec(nowUnixSec), pauseEndHour);
}

/**
 * Active (non-paused) seconds remaining from `nowUnixSec` until the pick expires
 * at `pickEndUnixSec`. This is what a slow-draft countdown should display: it
 * reads the full pick length at the start, ticks down only during active hours,
 * and freezes overnight.
 */
export function slowDraftActiveSecondsUntil(nowUnixSec: number, pickEndUnixSec: number, pauseEndHour: number = defaultPauseEndHour): number {
  return slowDraftEffectiveElapsedSeconds(nowUnixSec, pickEndUnixSec, pauseEndHour);
}

/** Seconds from `nowUnixSec` until tonight's 22:00 PT pause starts; 0 while paused. */
export function slowDraftSecondsUntilPause(nowUnixSec: number, pauseEndHour: number = defaultPauseEndHour): number {
  const w = nyWallFromUnixSec(nowUnixSec);
  if (inNightPause(w, pauseEndHour)) return 0;
  return Math.max(0, unixAtNYWallClock(w.y, w.mon, w.d, PAUSE_START_HOUR, 0, 0) - nowUnixSec);
}

/**
 * What a slow-draft clock should DISPLAY. Normally the active seconds left
 * (ticks in active hours, frozen overnight). With the fresh-clock-after-pause
 * rule a pick armed after (22:00 − pickLength) has MORE active time than one
 * clock: tonight's remainder PLUS a full clock at pauseEnd. Capping that at
 * pickLength (the pre-8/27 behavior) showed a flat "4:00:00" from ~6pm until
 * 7am — every user read it as frozen (ticket-2661). Richard 8/27: "it should
 * still count down." So while there is more than one clock ahead, count down
 * to the pause (tonight's real remaining), then the full clock shows through
 * the pause and ticks from pauseEnd. Never returns 0 before the pause: 0 flips
 * canDraft off in the room for the final second, so the floor is 1.
 */
export function slowDraftDisplaySecondsUntil(
  nowUnixSec: number,
  pickEndUnixSec: number,
  pickLengthSec: number,
  pauseEndHour: number = defaultPauseEndHour
): number {
  const active = slowDraftActiveSecondsUntil(nowUnixSec, pickEndUnixSec, pauseEndHour);
  if (!(pickLengthSec > 0) || active <= pickLengthSec) return active;
  const toPause = slowDraftSecondsUntilPause(nowUnixSec, pauseEndHour);
  return toPause > 0 ? Math.max(1, toPause) : pickLengthSec;
}

File: repos/sbs-drafts-api-deploy/models/slow_draft_clock.go
package models

import "time"

// Slow drafts only count pick time during the active window
// pauseEnd:00–22:00 America/Los_Angeles (22:00–pauseEnd is paused).
// pauseEnd is 05 legacy; system_config/slowDraftClock.pauseEndHour overrides
// it while the switch is active (Richard 2026-08-26: 7am).

var pacific *time.Location

const slowDraftPauseStartHour = 22

func init() {
	var err error
	pacific, err = time.LoadLocation("America/Los_Angeles")
	if err != nil {
		panic("slow draft clock: America/Los_Angeles: " + err.Error())
	}
}

func slowDraftInNightPauseAt(t time.Time, pauseEndHour int) bool {
	t = t.In(pacific)
	sod := t.Hour()*3600 + t.Minute()*60 + t.Second()
	return sod >= slowDraftPauseStartHour*3600 || sod < pauseEndHour*3600
}

func slowDraftInNightPause(t time.Time) bool {
	return slowDraftInNightPauseAt(t, SlowDraftPauseEndHour())
}

// nextPauseEnd returns pauseEnd:00 PT on the day after y/m/d.
func nextPauseEnd(y int, m time.Month, d int, pauseEndHour int) time.Time {
	midnight := time.Date(y, m, d, 0, 0, 0, 0, pacific)
	return midnight.AddDate(0, 0, 1).Add(time.Duration(pauseEndHour) * time.Hour)
}

// slowDraftAdvanceToNextActive returns the earliest instant >= t that is not in the night pause window.
func slowDraftAdvanceToNextActiveAt(t time.Time, pauseEndHour int) time.Time {
	t = t.In(pacific)
	if !slowDraftInNightPauseAt(t, pauseEndHour) {
		return t
	}
	y, m, d := t.Date()
	sod := t.Hour()*3600 + t.Minute()*60 + t.Second()
	if sod >= slowDraftPauseStartHour*3600 {
		return nextPauseEnd(y, m, d, pauseEndHour)
	}
	return time.Date(y, m, d, pauseEndHour, 0, 0, 0, pacific)
}

func slowDraftAdvanceToNextActive(t time.Time) time.Time {
	return slowDraftAdvanceToNextActiveAt(t, SlowDraftPauseEndHour())
}

// SlowDraftPickEndUnix returns the Unix instant when pickLengthSec of slow-draft clock have elapsed from fromUnix.
// Honours the system_config/slowDraftClock switch (fresh clock after the pause, pause end hour).
func SlowDraftPickEndUnix(fromUnix int64, pickLengthSec int64) int64 {
	return slowDraftPickEndUnixOpts(fromUnix, pickLengthSec, SlowDraftFreshClockAfterPause(), SlowDraftPauseEndHour())
}

// slowDraftPickEndUnixOpts is the pure form. freshAfterPause=true: a pick that
// would cross the 22:00 PT pause does NOT carry its leftover minutes into the
// morning — it restarts with the FULL pickLengthSec when the pause ends (so
// nobody wakes up to 40 minutes left). false: legacy carry-over.
func slowDraftPickEndUnixOpts(fromUnix int64, pickLengthSec int64, freshAfterPause bool, pauseEndHour int) int64 {
	if pickLengthSec <= 0 {
		return fromUnix
	}
	if pickLengthSec > int64(slowDraftPauseStartHour-pauseEndHour)*3600 {
		// Can't fit in one active window — a fresh clock would never end.
		freshAfterPause = false
	}
	cur := slowDraftAdvanceToNextActiveAt(time.Unix(fromUnix, 0), pauseEndHour)
	remaining := pickLengthSec
	for remaining > 0 {
		cur = cur.In(pacific)
		y, m, d := cur.Date()
		windowClose := time.Date(y, m, d, slowDraftPauseStartHour, 0, 0, 0, pacific)
		avail := int64(windowClose.Sub(cur).Seconds())
		if avail <= 0 {
			cur = nextPauseEnd(y, m, d, pauseEndHour)
			continue
		}
		if remaining <= avail {
			return cur.Add(time.Duration(remaining) * time.Second).Unix()
		}
		if freshAfterPause {
			remaining = pickLengthSec
		} else {
			remaining -= avail
		}
		cur = nextPauseEnd(y, m, d, pauseEndHour)
	}
	return cur.Unix()
}

// SlowDraftEffectiveElapsedSeconds returns how many slow-draft "active" seconds elapse between startUnix and endUnix.
func SlowDraftEffectiveElapsedSeconds(startUnix, endUnix int64) int64 {
	return slowDraftEffectiveElapsedSecondsAt(startUnix, endUnix, SlowDraftPauseEndHour())
}

func slowDraftEffectiveElapsedSecondsAt(startUnix, endUnix int64, pauseEndHour int) int64 {
	if endUnix <= startUnix {
		return 0
	}
	var total int64
	cur := slowDraftAdvanceToNextActiveAt(time.Unix(startUnix, 0), pauseEndHour)
	endT := time.Unix(endUnix, 0).In(pacific)
	for cur.Before(endT) {
		cur = cur.In(pacific)
		y, m, d := cur.Date()
		windowClose := time.Date(y, m, d, slowDraftPauseStartHour, 0, 0, 0, pacific)
		if !windowClose.After(cur) {
			cur = nextPauseEnd(y, m, d, pauseEndHour)
			continue
		}
		chunkEnd := windowClose
		if endT.Before(chunkEnd) {
			chunkEnd = endT
		}
		total += int64(chunkEnd.Sub(cur).Seconds())
		if !chunkEnd.Before(endT) {
			break
		}
		cur = nextPauseEnd(y, m, d, pauseEndHour)
	}
	return total
}
